dask/array/routines.py

```diff
@@ -1190,3 +1190,101 @@ def insert(arr, obj, values, axis):
     interleaved = list(interleave([split_arr, split_values]))
     interleaved = [i for i in interleaved if i.nbytes]
     return concatenate(interleaved, axis=axis)
+
+
+def _einsum_kernel_nooptimize(*operands, **kwargs):
+    chunk = np.einsum(kwargs['subscripts'], *operands,
+                      casting=kwargs['casting'],
+                      dtype=kwargs['kernel_dtype'],
+                      order=kwargs['order'])
+
+    # Avoid concatenate=True in atop by adding 1's
+    # for the contracted dimensions
+    return chunk.reshape(chunk.shape + (1,) * kwargs['ncontract_inds'])
+
+
+def _einsum_kernel_optimize(*operands, **kwargs):
+    chunk = np.einsum(kwargs['subscripts'], *operands,
+                      casting=kwargs['casting'],
+                      dtype=kwargs['kernel_dtype'],
+                      order=kwargs['order'],
+                      optimize=kwargs['optimize'])
```

Comment:
Might be able to `pop` a few values from `kwargs` and pass the rest through. That could also smooth over the differences between the supported optimize case and non-optimize/unsupported case. Thus simplifying the surrounding code.

Reply:
+1 for using `pop`, e.g.,
```
def _einsum_kernel(*operands, **kwargs):
    subscripts = kwargs.pop('subscripts')
    ncontract_inds = kwargs.pop('ncontract_inds')
    chunk = np.einsum(subsripts, *operands, **kwargs)
    ...
```